felix/fv/bpf_dual_stack_test.go

```diff
@@ -0,0 +1,540 @@
+				},
+			}
+			pol.Spec.Egress = []api.Rule{
+				{
+					Action: "Allow",
+					Source: api.EntityRule{
+						Selector: "workload=='regular'",
+					},
+				},
+			}
+			pol.Spec.Selector = "workload=='regular'"
+
+			pol = createPolicy(pol)
+		})
+
+		AfterEach(func() {
+			if CurrentGinkgoTestDescription().Failed {
+				infra.DumpErrorData()
+				for _, felix := range tc.Felixes {
+					felix.Exec("calico-bpf", "counters", "dump")
+				}
+			}
+
+			for i := 0; i < 2; i++ {
+				for j := 0; j < 2; j++ {
+					w[i][j].Stop()
+				}
+			}
+			_, err := calicoClient.GlobalNetworkPolicies().Delete(context.Background(), "policy-1", options2.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			tc.Stop()
+			infra.Stop()
+		})
+
+		Context("with a policy allowing ingress to w[0][0] from all regular workloads", func() {
+			var (
+				testSvc          *v1.Service
+				testSvcNamespace string
+			)
+
+			npPort := uint16(30333)
+			clusterIPs := []string{"10.101.0.10", "dead:beef::abcd:0:0:10"}
+			BeforeEach(func() {
+				k8sClient = infra.(*infrastructure.K8sDatastoreInfra).K8sClient
+				_ = k8sClient
+				testSvc = k8sServiceForDualStack("test-svc", clusterIPs, w[0][0], 80, 8055, int32(npPort), "tcp")
+				testSvcNamespace = testSvc.ObjectMeta.Namespace
+				_, err := k8sClient.CoreV1().Services(testSvcNamespace).Create(context.Background(), testSvc, metav1.CreateOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				Eventually(k8sGetEpsForServiceFunc(k8sClient, testSvc), "10s").Should(HaveLen(1),
+					"Service endpoints didn't get created? Is controller-manager happy?")
+			})
+			It("Should connect to w[0][0] from all other workloads with IPv4 and IPv6", func() {
+				cc.ExpectSome(w[0][1], w[0][0])
+				cc.ExpectSome(w[1][0], w[0][0])
+				cc.ExpectSome(w[1][1], w[0][0])
+
+				cc.Expect(Some, w[0][1], w[0][0], ExpectWithIPVersion(6))
+				cc.Expect(Some, w[1][0], w[0][0], ExpectWithIPVersion(6))
+				cc.Expect(Some, w[1][1], w[0][0], ExpectWithIPVersion(6))
+				cc.CheckConnectivity()
+			})
+
+			It("Should connect to w[0][0] via clusterIP (IPv4 and IPv6)", func() {
+				port := uint16(testSvc.Spec.Ports[0].Port)
+				cc.ExpectSome(w[1][0], TargetIP(clusterIPs[0]), port)
+				cc.ExpectSome(w[1][0], TargetIP(clusterIPs[1]), port)
+
+				cc.ExpectSome(w[0][1], TargetIP(clusterIPs[0]), port)
+				cc.ExpectSome(w[0][1], TargetIP(clusterIPs[1]), port)
+				cc.CheckConnectivity()
+			})
+
+			It("Should connect to w[0][0] via nodePort (IPv4 and IPv6)", func() {
+				cc.ExpectSome(w[1][0], TargetIP(felixIP(0)), npPort)
+				cc.ExpectSome(w[0][1], TargetIP(felixIP(0)), npPort)
+
+				cc.ExpectSome(w[1][0], TargetIP(felixIP6(0)), npPort)
+				cc.ExpectSome(w[0][1], TargetIP(felixIP6(0)), npPort)
+				cc.CheckConnectivity()
+			})
+
+		})
+
+	})
+}
+
+var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ BPF dual stack tests with delayed start", []apiconfig.DatastoreType{apiconfig.Kubernetes}, func(getInfra infrastructure.InfraFactory) {
```

Comment:
does he "with delayed start" say much about what is being tested?

Why not to merge this the test above to save yourself all the copy-pasta of the boilerplate code? If you want to run the above test with ctlb on/off you can have it in a function which you then call 2x from IT. The tests without delay can have the trigger in the "justbeforeeach" or you can setup the infra i "justbeforeeach" and have a setting switch in "justbefore" whether it should delay or not.